smartsim/_core/entrypoints/indirect.py

```diff
@@ -0,0 +1,237 @@
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import logging
+import os
+import pathlib
+import signal
+import sys
+import typing as t
+from types import FrameType
+
+import coloredlogs
+import psutil
+
+import smartsim.log
+from smartsim._core.entrypoints.telemetrymonitor import track_event
+from smartsim._core.utils.helpers import decode_cmd, get_ts
+
+STEP_PID: t.Optional[int] = None
+logger = smartsim.log.get_logger(__name__)
+
+# kill is not catchable
+SIGNALS = [signal.SIGINT, signal.SIGTERM, signal.SIGQUIT, signal.SIGABRT]
+
+
+def main(
+    cmd: str,
+    etype: str,
+    cwd: str,
+    status_dir: str,
+) -> int:
+    """Execute the step command and emit tracking events"""
+    global STEP_PID  # pylint: disable=global-statement
+    proxy_pid = os.getpid()
+
+    status_path = pathlib.Path(status_dir)
+    if not status_path.exists():
+        status_path.mkdir(parents=True, exist_ok=True)
+
+    if not cmd.strip():
+        raise ValueError("Invalid cmd supplied")
+
+    cleaned_cmd = decode_cmd(cmd)
+    ret_code: int = 1
+    logger.debug("Indirect step starting")
+
+    start_detail = f"Proxy process {proxy_pid}"
+    start_rc: t.Optional[int] = None
+
+    try:
+        process = psutil.Popen(
+            cleaned_cmd,
+            cwd=cwd,
+            stdout=sys.stdout,
+            stderr=sys.stderr,
+        )
+        STEP_PID = process.pid
+        logger.info(f"Indirect proxy {proxy_pid} child process {STEP_PID} started")
+        start_detail += f" started child process {STEP_PID}"
+
+    except Exception as ex:
+        start_detail += f" failed to start child process. {ex}"
+        start_rc = 1
+        logger.error("Failed to create process", exc_info=True)
+        cleanup()
+        return 1
+    finally:
+        track_event(
+            get_ts(),
+            proxy_pid,
+            "",  # step_id for unmanaged task is always empty
+            etype,
+            "start",
+            status_path,
+            logger,
+            detail=start_detail,
+            return_code=start_rc,
+        )
```

Comment:
Small fun bug here for the validate CLI:

If a user were to run ``smart validate`` with the TM enabled, this entry-point process could live on beyond the time the validation experiment has concluded. This means that this ``track_event`` call may actually recreate the temporary directory that the validation driver was using as an experiment directory in order to write out the ``stop.json`` file (this was at least consistently the case for me when I ran ``smart validate``).

There is nothing "wrong" with this, but it would be a nice QoL feature if the ``validate`` command did not leave artifacts around if the validation was successful. The easiest ways I can think of to guarantee this would be to either:

1) Automatically turn the TM off if a user runs ``smart validate``, regardless of the ``SMARTSIM_FLAG_TELEMETRY`` env var (this is my preferred solution).
1) Validate that all TM and indirect processes have finished before the ``validate`` driver cleans up the temp exp dir.

Reply:
Added `_disable_telmon` context manager to disable telemetry monitor for the test.